### ISSUE TITLE
Support from Pluto.jl

### ISSUE THREAD

**Issue by ahjulstad** [JuliaDynamics/ResumableFunctions.jl#111]
Would it be possible run inside of Pluto.jl? Initial testing with ConcurrentSim.jl (and ResumableFunctions.jl) give errors in Pluto.jl. 

Just something along the lines of the example

```
using Distributions, ResumableFunctions, ConcurrentSim, Random

@resumable function fibonacci(n::Int)
  a = 0
  b = 1
  for i in 1:n
    @yield a
    a, b = b, a+b
  end
end

for val in fibonacci(10) 
  println(val) 
end
```

Gives error message

```
UndefVarError: `##236` not defined in `Main.var"workspace#6"`
Suggestion: check for spelling errors or missing imports.
```
at the last line (the end statement) of the fibonacci function. 

I have no idea how difficult this is, but I think the interactivity of Pluto together with the speed of simulations in Julia can be useful.

**Comment by thofma** [comment]
Since this works everywhere else, this looks like a Pluto.jl bug.

**Comment by fonsp** [comment]
Looks like it's not working because part of the code is already evaluated during macro expansion, here:

https://github.com/JuliaDynamics/ResumableFunctions.jl/blob/932dca3e0b69cf42db0d8db12b599ea0a9b9075e/src/utils.jl#L78C9-L78C12

Pluto supports regular macros, but side effects from macro expansion are too difficult to support.

A solution could be to return `func_expr` from `get_slots` instead of the direct `@eval`, and insert it in the returned expression [here](https://github.com/JuliaDynamics/ResumableFunctions.jl/blob/932dca3e0b69cf42db0d8db12b599ea0a9b9075e/src/macro.jl#L218).

**Comment by Krastanov** [comment]
Thanks, @fonsp, for looking into this! I think it is slightly more complicated than that. This library also needs to run inference on `func_expr` (couple of lines below)

```
  @eval(mod, @noinline $func_expr)
  #@info func_def[:body]|>MacroTools.striplines
  # get typed code
  codeinfos = @eval(mod, code_typed($(func_def[:name]), Tuple; optimize=false))
```

I do not really know how to get `codeinfos` without the side effect of the eval. Is there some sandboxed way to do this without a side effect?

**Comment by fonsp** [comment]
Ah I see! Sorry I didn't see that! Indeed it makes sense to do the eval like you do now :)

In that case, it's just about tweaking the macro hygiene: you could change this:

https://github.com/JuliaDynamics/ResumableFunctions.jl/blob/6d9ce3b3c05286d6efe92c3faf751ed4659f4f47/src/utils.jl#L97

into:

```julia
  inferfn_reference = :($(mod).$(func_def[:name]))
  return inferfn_reference, slots
```

to refer to the function by its full name (Main.Wherever.var"#12") instead of just the generated Symbol name (var"#12#13").

Or maybe this (store the function by reference, not by name): https://github.com/JuliaDynamics/ResumableFunctions.jl/compare/master...fonsp:ResumableFunctions.jl:patch-1

**Comment by Krastanov** [comment]
Thank you, much appreciated!

**Comment by Krastanov** [comment]
fixed with fonsp's patch in #117 

new release to be made shortly

@ahjulstad , let us know if there are continuing problems
